README.md
arc-dataquality-udf-plugin defines a set of data quality/validation user defined functions.

The intention of this plugin is to provide a standard suite of data quality functions that can be used by many people and companies. Feel free to raise pull requests against this repository with reusable functions that may be beneficial to others.

## User Defined Functions

- `is_valid_phonenumber(numberToParse, defaultRegion)` returns a boolean of whether the phone number appears valid for that region e.g. `is_valid_phonenumber('61499000000', 'AU')`.
- `format_phonenumber(numberToParse, defaultRegion)` returns an [ISO E164](https://en.wikipedia.org/wiki/E.164) formatted phone number string e.g. `format_phonenumber('61499000000', 'AU')` becomes `+61499000000`.
- `is_valid_abn(abn)` returns a boolean of whether the Australian Business Number (ABN) passes the inbuilt checksum function.
- `is_valid_acn(acn)` returns a boolean of whether the Australian Company Number (ACN) passes the inbuilt checksum function.

## License

Arc is released under the [MIT License](https://opensource.org/licenses/MIT).

`is_valid_phonenumber` and `format_phonenumber` use https://github.com/google/libphonenumber released under the [Apache 2.0](https://opensource.org/licenses/Apache-2.0) license.

CHANGELOG.md
# Change Log

## 1.8.0

- bump to Arc 3.13.2

## 1.7.0

- bump to Arc 3.7.0
- bump to libphonenumber 8.12.15

## 1.6.0

- bump to Arc 3.4.0
- bump to libphonenumber 8.12.10

## 1.5.0

- bump to Arc 3.2.0
- bump to libphonenumber 8.12.9

## 1.4.0

- bump to Arc 3.0.0
- bump to libphonenumber 8.12.5

## 1.3.0

- bump to Arc 2.10.0
- bump to libphonenumber 8.12.1

**NOTE** This is the last release supporting `Scala 2.11` given the release of `Spark 3.0` which only supports `Scala 2.12`.

## 1.2.0

- bump to Arc 2.8.0
- bump to libphonenumber 8.12.0

## 1.1.0

- is_valid_acn(acn)

## 1.0.0

- initial release.
- is_valid_phonenumber(numberToParse, defaultRegion)
- format_phonenumber(numberToParse, defaultRegion)
- is_valid_abn(abn)


src/test/scala/ai/tripl/arc/util/TestUtils.scala
package ai.tripl.arc.util

import java.io.File
import java.sql.Date
import java.sql.Timestamp
import java.time.{ZoneId, ZonedDateTime}
import java.util.ServiceLoader
import scala.collection.JavaConverters._

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.core._

import com.typesafe.config.ConfigFactory

import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.rdd.RDD
import org.apache.spark.storage.StorageLevel

import ai.tripl.arc.api.API.ARCContext
import ai.tripl.arc.plugins._
import ai.tripl.arc.util.log.LoggerFactory
import org.apache.log4j.{Level, Logger}

case class KnownData(
    booleanDatum: Boolean,
    dateDatum: Date,
    decimalDatum: Decimal,
    doubleDatum: Double,
    integerDatum: Integer,
    longDatum: Long,
    stringDatum: String,
    timeDatum: String,
    timestampDatum: Timestamp,
    nullDatum: Null
)

object TestUtils {
    def getLogger()(implicit spark: SparkSession): ai.tripl.arc.util.log.logger.Logger = {
        val loader = ai.tripl.arc.util.Utils.getContextOrSparkClassLoader
        val logger = LoggerFactory.getLogger(spark.sparkContext.applicationId)
        Logger.getLogger("org").setLevel(Level.ERROR)
        Logger.getLogger("breeze").setLevel(Level.ERROR)
        logger
    }

    def getARCContext(
        isStreaming: Boolean = false,
        ignoreEnvironments: Boolean = false,
        immutableViews: Boolean = false,
        environment: String = "test",
        commandLineArguments: Map[String,String] = Map[String,String](),
        ipynb: Boolean = true,
        inlineSQL: Boolean = true,
        inlineSchema: Boolean = true,
        dropUnsupported: Boolean = false,
        completionEnvironments: List[String] = List("production", "test")
    )(implicit spark: SparkSession): ARCContext = {
      val loader = ai.tripl.arc.util.Utils.getContextOrSparkClassLoader

      ARCContext(
        jobId=None,
        jobName=None,
        environment=Option(environment),
        configUri=None,
        isStreaming=isStreaming,
        ignoreEnvironments=ignoreEnvironments,
        commandLineArguments=commandLineArguments,
        storageLevel=StorageLevel.MEMORY_AND_DISK_SER,
        immutableViews=immutableViews,
        ipynb=ipynb,
        inlineSQL=inlineSQL,
        inlineSchema=inlineSchema,
        dropUnsupported=dropUnsupported,
        dynamicConfigurationPlugins=ServiceLoader.load(classOf[DynamicConfigurationPlugin], loader).iterator().asScala.toList,
        lifecyclePlugins=ServiceLoader.load(classOf[LifecyclePlugin], loader).iterator().asScala.toList,
        activeLifecyclePlugins=Nil,
        pipelineStagePlugins=ServiceLoader.load(classOf[PipelineStagePlugin], loader).iterator().asScala.toList,
        udfPlugins=ServiceLoader.load(classOf[UDFPlugin], loader).iterator().asScala.toList,
        serializableConfiguration=new SerializableConfiguration(spark.sparkContext.hadoopConfiguration),
        userData=collection.mutable.Map.empty,
        resolutionConfig=ConfigFactory.load(),
        completionEnvironments=completionEnvironments
      )
    }

    def datasetEquality(expected: DataFrame, actual: DataFrame)(implicit spark: SparkSession): Boolean = {
        import spark.implicits._

        // if both are empty ignore
        if (expected.count != 0 || actual.count != 0) {

            val expectedHashDF = expected.withColumn("_hashLeft", sha2(to_json(struct(expected.columns.sorted.map(col):_*)),512))
            val actualHashDF = actual.withColumn("_hashRight", sha2(to_json(struct(actual.columns.sorted.map(col):_*)),512))

            val transformedDF = expectedHashDF
                .joinWith(actualHashDF, expectedHashDF("_hashLeft") === actualHashDF("_hashRight"), "full")
                .withColumnRenamed("_1", "expected")
                .withColumnRenamed("_2", "actual")

            transformedDF.persist

            val expectedExceptActual = transformedDF.filter(col("actual").isNull)
            val actualExceptExpected = transformedDF.filter(col("expected").isNull)
            val expectedExceptActualCount = expectedExceptActual.count
            val actualExceptExpectedCount = actualExceptExpected.count

            if (expectedExceptActualCount != 0 || actualExceptExpectedCount != 0) {
                println("EXPECTED")
                println(expected.schema)
                expected.show(false)
                println("ACTUAL")
                println(actual.schema)
                actual.show(false)

                transformedDF.unpersist
                false
            } else {
                true
            }
        } else {
            true
        }
    }

    def getKnownDataset()(implicit spark: SparkSession): DataFrame = {
        import spark.implicits._

        val dataset = Seq(
            KnownData(booleanDatum=true, dateDatum=Date.valueOf("2016-12-18"), decimalDatum=Decimal(54.321, 10, 3), doubleDatum=42.4242, integerDatum=17, longDatum=1520828868, stringDatum="test,breakdelimiter", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 20, 21, 46, 54, 0, ZoneId.of("UTC")).toInstant), timeDatum="12:34:56", nullDatum=null),
            KnownData(booleanDatum=false, dateDatum=Date.valueOf("2016-12-19"), decimalDatum=Decimal(12.345, 10, 3), doubleDatum=21.2121, integerDatum=34, longDatum=1520828123, stringDatum="breakdelimiter,test", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 29, 17, 21, 49, 0, ZoneId.of("UTC")).toInstant), timeDatum="23:45:16", nullDatum=null)
        )

        dataset.toDF
    }

    // modified dataset for DiffTransform test
    def getKnownAlteredDataset()(implicit spark: SparkSession): DataFrame = {
        import spark.implicits._

        val dataset = Seq(
            // same first row
            KnownData(booleanDatum=true, dateDatum=Date.valueOf("2016-12-18"), decimalDatum=Decimal(54.321, 10, 3), doubleDatum=42.4242, integerDatum=17, longDatum=1520828868, stringDatum="test,breakdelimiter", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 20, 21, 46, 54, 0, ZoneId.of("UTC")).toInstant), timeDatum="12:34:56", nullDatum=null),
            // altered second row (only booleanDatum and integerValue value has been changed)
            KnownData(booleanDatum=true, dateDatum=Date.valueOf("2016-12-19"), decimalDatum=Decimal(12.345, 10, 3), doubleDatum=21.2121, integerDatum=35, longDatum=1520828123, stringDatum="breakdelimiter,test", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 29, 17, 21, 49, 0, ZoneId.of("UTC")).toInstant), timeDatum="23:45:16", nullDatum=null)
        )

        dataset.toDF
    }

    def getKnownStringDataset()(implicit spark: SparkSession): DataFrame = {
        val df = getKnownDataset()
        df.select(df.columns.map(c => col(c).cast(StringType)) : _*)
    }

    def knownDatasetPrettyJSON(row: Int)(implicit spark: SparkSession): String = {
        val json = getKnownDataset().toJSON.collect.toList(row)
        val objectMapper = new ObjectMapper()
        val jsonTree = objectMapper.readTree(json)

        objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonTree)
    }

    def getListOfFiles(dir: String):List[File] = {
        val d = new File(dir)
        if (d.exists && d.isDirectory) {
            d.listFiles.filter(_.isFile).filter(f => !f.getPath.contains("README.md")).toList
        } else {
            List[File]()
        }
    }

    def getKnownDatasetMetadataJson(): String = {
    """
    [
        {
            "id": "982cbf60-7ba7-4e50-a09b-d8624a5c49e6",
            "name": "booleanDatum",
            "description": "booleanDatum",
            "type": "boolean",
            "trim": false,
            "nullable": false,
            "nullableValues": [
                "",
                "null"
            ],
            "trueValues": [
                "true"
            ],
            "falseValues": [
                "false"
            ],
            "metadata": {
                "booleanMeta": true,
                "booleanArrayMeta": [true, false],
                "stringMeta": "string",
                "stringArrayMeta": ["string0", "string1"],
                "longMeta": 10,
                "longArrayMeta": [10,20],
                "doubleMeta": 0.141,
                "doubleArrayMeta": [0.141, 0.52],
                "private": false,
                "securityLevel": 0
            }
        },
        {
            "id": "0e8109ba-1000-4b7d-8a4c-b01bae07027f",
            "name": "dateDatum",
            "description": "dateDatum",
            "type": "date",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "formatters": [
                "yyyy-MM-dd"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 3
            }
        },
        {
            "id": "9712c383-22d1-44a6-9ca2-0087af4857f1",
            "name": "decimalDatum",
            "description": "decimalDatum",
            "type": "decimal",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "precision": 38,
            "scale": 18,
            "metadata": {
                "private": true,
                "securityLevel": 2
            }
        },
        {
            "id": "31541ea3-5b74-4753-857c-770bd601c35b",
            "name": "doubleDatum",
            "description": "doubleDatum",
            "type": "double",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 8
            }
        },
        {
            "id": "a66f3bbe-d1c6-44c7-b096-a4be59fdcd78",
            "name": "integerDatum",
            "description": "integerDatum",
            "type": "integer",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 10
            }
        },
        {
            "id": "1c0eec1d-17cd-45da-8744-7a9ef5b8b086",
            "name": "longDatum",
            "description": "longDatum",
            "type": "long",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": false,
                "securityLevel": 0
            }
        },
        {
            "id": "9712c383-22d1-44a6-9ca2-0087af4857f1",
            "name": "stringDatum",
            "description": "stringDatum",
            "type": "string",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": false,
                "securityLevel": 0
            }
        },
        {
            "id": "eb17a18e-4664-4016-8beb-cd2a492d4f20",
            "name": "timeDatum",
            "description": "timeDatum",
            "type": "time",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "formatters": [
                "HH:mm:ss"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 8
            }
        },
        {
            "id": "8e42c8f0-22a8-40db-9798-6dd533c1de36",
            "name": "timestampDatum",
            "description": "timestampDatum",
            "type": "timestamp",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "formatters": [
                "uuuu-MM-dd'T'HH:mm:ss.SSSXXX"
            ],
            "timezoneId": "UTC",
            "metadata": {
                "private": true,
                "securityLevel": 7
            }
        }
    ]
    """
    }
}



src/test/scala/ai/tripl/arc/plugins/udf/DataQualitySuite.scala
package ai.tripl.arc.plugins

import ai.tripl.arc.ARC
import ai.tripl.arc.api.API._
import ai.tripl.arc.config.ArcPipeline
import ai.tripl.arc.config.Error._
import ai.tripl.arc.util.TestUtils

import org.apache.spark.sql.SparkSession
import org.scalatest.{BeforeAndAfter, FunSuite}
import ai.tripl.arc.udf.UDF


class DataQualitySuite extends FunSuite with BeforeAndAfter {

  var session: SparkSession = _  

  before {
    implicit val spark = SparkSession
                  .builder()
                  .master("local[*]")
                  .config("spark.ui.port", "9999")
                  .appName("Spark ETL Test")
                  .getOrCreate()
    spark.sparkContext.setLogLevel("INFO")
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    // set for deterministic timezone
    spark.conf.set("spark.sql.session.timeZone", "UTC")   

    UDF.registerUDFs()(spark, logger, arcContext)

    session = spark
    import spark.implicits._
  }

  after {
    session.stop()
  }

  test("DataQualitySuite") {
    implicit val spark = session
    import spark.implicits._
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    val conf = s"""{
      "plugins": {
        "config": []
      },
      "stages": [
        {
          "type": "SQLTransform",
          "name": "test",
          "description": "test",
          "environments": [
            "production",
            "test"
          ],
          "inputURI": "${spark.getClass.getResource("/").toString}/basic.sql",
          "outputView": "customer",
          "persist": false,
        }
      ]
    }"""
    
    val pipelineEither = ArcPipeline.parseConfig(Left(conf), arcContext)

    // assert graph created
    pipelineEither match {
      case Left(err) => fail(err.toString)
      case Right((pipeline, _)) => {
        val row = ARC.run(pipeline)(spark, logger, arcContext).get.first
        
        assert(row.getBoolean(0) == true)
        assert(row.getBoolean(1) == false)
        assert(row.getString(2) == "+61499000000")
        assert(row.getBoolean(3) == true)
        assert(row.getBoolean(4) == false)
        assert(row.getBoolean(5) == true)
        assert(row.getBoolean(6) == false)        
      }
    }
  }
  
  test("isValidPhoneNumber") {
    case class PhoneNumberValid(
      number: String,
      region: String,
      valid: Boolean
    )

    val tests = List(
      PhoneNumberValid("61499000000", "AU", true),
      PhoneNumberValid("0499000000", "AU", true),
      PhoneNumberValid("61499000000", "US", false),
      PhoneNumberValid("888-452-1505", "US", true)
    )

    tests.foreach{ 
      t: PhoneNumberValid => {
        assert(ai.tripl.arc.plugins.udf.DataQualityPlugin.isValidPhoneNumber(t.number, t.region) == t.valid, s"${t.number} ${t.region}")
      }
    }
  }

  test("formatPhoneNumber") {
    case class PhoneNumberValid(
      number: String,
      region: String,
      valid: Option[String]
    )

    val tests = List(
      PhoneNumberValid("61499000000", "AU", Option("+61499000000")),
      PhoneNumberValid("(555) 555-1234", "US", None),
      PhoneNumberValid("61499000000", "US", None)
    )

    tests.foreach{ 
      t: PhoneNumberValid => {
        t.valid match {
          case Some(valid) => assert(ai.tripl.arc.plugins.udf.DataQualityPlugin.formatPhoneNumber(t.number, t.region) == valid, s"${t.number} ${t.region}")
          case None => {
            val thrown0 = intercept[Exception] {
              ai.tripl.arc.plugins.udf.DataQualityPlugin.formatPhoneNumber(t.number, t.region)
            }
            assert(thrown0.getMessage === "Cannot format invalid phone number.")
          }
        }
        
      }
    }
  }  

  test("isValidABN") {
    case class ABNValid(
      abn: String,
      valid: Boolean
    )

    val tests = List(
      ABNValid("83 914 571 673", true),
      ABNValid("83 914 571 672", false),
      ABNValid("83 914 571 67", false),
      ABNValid("83 914 571 67a", false)
    )

    tests.foreach{ 
      t: ABNValid => {
        assert(ai.tripl.arc.plugins.udf.DataQualityPlugin.isValidABN(t.abn) == t.valid, s"${t.abn}")
      }
    }
  }  

  test("isValidACN") {
    case class ACNValid(
      acn: String,
      valid: Boolean
    )

    val tests = List(
      ACNValid("000 000 019", true),
      ACNValid("000 250 000", true),
      ACNValid("000 500 005", true),
      ACNValid("000 750 005", true),
      ACNValid("001 000 004", true),
      ACNValid("001 250 004", true),
      ACNValid("001 500 009", true),
      ACNValid("001 749 999", true),
      ACNValid("001 999 999", true),
      ACNValid("002 249 998", true),
      ACNValid("002 499 998", true),
      ACNValid("002 749 993", true),
      ACNValid("002 999 993", true),
      ACNValid("003 249 992", true),
      ACNValid("003 499 992", true),
      ACNValid("003 749 988", true),
      ACNValid("003 999 988", true),
      ACNValid("004 249 987", true),
      ACNValid("004 499 987", true),
      ACNValid("004 749 982", true),
      ACNValid("004 999 982", true),
      ACNValid("005 249 981", true),
      ACNValid("005 499 981", true),
      ACNValid("005 749 986", true),
      ACNValid("005 999 977", true),
      ACNValid("006 249 976", true),
      ACNValid("006 499 976", true),
      ACNValid("006 749 980", true),
      ACNValid("006 999 980", true),
      ACNValid("007 249 989", true),
      ACNValid("007 499 989", true),
      ACNValid("007 749 975", true),
      ACNValid("007 999 975", true),
      ACNValid("008 249 974", true),
      ACNValid("008 499 974", true),
      ACNValid("008 749 979", true),
      ACNValid("008 999 979", true),
      ACNValid("009 249 969", true),
      ACNValid("009 499 969", true),
      ACNValid("009 749 964", true),
      ACNValid("009 999 964", true),
      ACNValid("010 249 966", true),
      ACNValid("010 499 966", true),
      ACNValid("010 749 961", true),
      ACNValid("000 000 010", false),
      ACNValid("000 250 001", false),
      ACNValid("000 500 006", false),
      ACNValid("000 750 006", false),
      ACNValid("001 000 005", false),
      ACNValid("001 250 005", false),
      ACNValid("001 500 000", false),
      ACNValid("001 749 990", false),
      ACNValid("001 999 990", false),
      ACNValid("002 249 999", false),
      ACNValid("002 499 999", false),
      ACNValid("002 749 994", false),
      ACNValid("002 999 994", false),
      ACNValid("003 249 993", false),
      ACNValid("003 499 993", false),
      ACNValid("003 749 989", false),
      ACNValid("003 999 989", false),
      ACNValid("004 249 988", false),
      ACNValid("004 499 988", false),
      ACNValid("004 749 983", false),
      ACNValid("004 999 983", false),
      ACNValid("005 249 982", false),
      ACNValid("005 499 982", false),
      ACNValid("005 749 987", false),
      ACNValid("005 999 978", false),
      ACNValid("006 249 977", false),
      ACNValid("006 499 977", false),
      ACNValid("006 749 981", false),
      ACNValid("006 999 981", false),
      ACNValid("007 249 980", false),
      ACNValid("007 499 980", false),
      ACNValid("007 749 976", false),
      ACNValid("007 999 976", false),
      ACNValid("008 249 975", false),
      ACNValid("008 499 975", false),
      ACNValid("008 749 970", false),
      ACNValid("008 999 970", false),
      ACNValid("009 249 960", false),
      ACNValid("009 499 960", false),
      ACNValid("009 749 965", false),
      ACNValid("009 999 965", false),
      ACNValid("010 249 967", false),
      ACNValid("010 499 967", false),
      ACNValid("010 749 962", false),     
      ACNValid("010 749 96", false),     
      ACNValid("010 749 96a", false)     
    )

    tests.foreach{ 
      t: ACNValid => {
        assert(ai.tripl.arc.plugins.udf.DataQualityPlugin.isValidACN(t.acn) == t.valid, s"${t.acn}")
      }
    }
  }    
}


src/test/scala/ai/tripl/arc/plugins/udf/DataValidationPluginSuite.scala
package ai.tripl.arc.plugins

import ai.tripl.arc.ARC
import ai.tripl.arc.api.API._
import ai.tripl.arc.config.ArcPipeline
import ai.tripl.arc.config.Error._
import ai.tripl.arc.util.TestUtils

import org.apache.spark.sql.SparkSession
import org.scalatest.{BeforeAndAfter, FunSuite}
import ai.tripl.arc.udf.UDF


class DataValidationPluginSuite extends FunSuite with BeforeAndAfter {

  var session: SparkSession = _  

  before {
    implicit val spark = SparkSession
                  .builder()
                  .master("local[*]")
                  .config("spark.ui.port", "9999")
                  .appName("Spark ETL Test")
                  .getOrCreate()
    spark.sparkContext.setLogLevel("INFO")
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    // set for deterministic timezone
    spark.conf.set("spark.sql.session.timeZone", "UTC")   

    UDF.registerUDFs()(spark, logger, arcContext)

    session = spark
    import spark.implicits._
  }

  after {
    session.stop()
  }

  test("DataValidationPluginSuite") {
    implicit val spark = session
    import spark.implicits._
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    val conf = s"""{
      "plugins": {
        "config": []
      },
      "stages": [
        {
          "type": "SQLTransform",
          "name": "test",
          "description": "test",
          "environments": [
            "production",
            "test"
          ],
          "inputURI": "${spark.getClass.getResource("/").toString}/basic.sql",
          "outputView": "customer",
          "persist": false,
        }
      ]
    }"""
    
    val pipelineEither = ArcPipeline.parseConfig(Left(conf), arcContext)

    // assert graph created
    pipelineEither match {
      case Left(err) => {
        println(err)
        assert(false)
      }
      case Right((pipeline, _)) => {
        val row = ARC.run(pipeline)(spark, logger, arcContext).get.first
        
        assert(row.getBoolean(0) == true)
        assert(row.getBoolean(1) == false)
        assert(row.getString(2) == "+61499000000")
        assert(row.getBoolean(3) == true)
        assert(row.getBoolean(4) == false)
      }
    }
  }
  
  test("isValidPhoneNumber") {
    case class PhoneNumberValid(
      number: String,
      region: String,
      valid: Boolean,
    )

    val tests = List(
      PhoneNumberValid("61499000000", "AU", true),
      PhoneNumberValid("0499000000", "AU", true),
      PhoneNumberValid("61499000000", "US", false),
      PhoneNumberValid("888-452-1505", "US", true)
    )

    tests.foreach{ 
      t: PhoneNumberValid => {
        assert(ai.tripl.arc.plugins.udf.DataValidationPlugin.isValidPhoneNumber(t.number, t.region) == t.valid, s"${t.number} ${t.region}")
      }
    }
  }

  test("formatPhoneNumber") {
    case class PhoneNumberValid(
      number: String,
      region: String,
      valid: String,
    )

    val tests = List(
      PhoneNumberValid("61499000000", "AU", "+61499000000"),
      PhoneNumberValid("(555) 555-1234", "US", "+15555551234")
    )

    tests.foreach{ 
      t: PhoneNumberValid => {
        assert(ai.tripl.arc.plugins.udf.DataValidationPlugin.formatPhoneNumber(t.number, t.region) == t.valid, s"${t.number} ${t.region}")
      }
    }
  }  

  test("isValidABN") {
    case class ABNValid(
      abn: String,
      valid: Boolean,
    )

    val tests = List(
      ABNValid("83 914 571 673", true),
      ABNValid("83 914 571 672", false)
    )

    tests.foreach{ 
      t: ABNValid => {
        assert(ai.tripl.arc.plugins.udf.DataValidationPlugin.isValidABN(t.abn) == t.valid, s"${t.abn}")
      }
    }
  }  
}


src/main/scala/ai/tripl/arc/plugins/udf/DataValidationPlugin.scala
package ai.tripl.arc.plugins.udf

import scala.collection.JavaConverters._

import org.apache.spark.sql.SparkSession

import ai.tripl.arc.util.log.logger.Logger
import ai.tripl.arc.api.API.ARCContext
import ai.tripl.arc.util.Utils

import com.google.i18n.phonenumbers.PhoneNumberUtil
import com.google.i18n.phonenumbers.PhoneNumberUtil.PhoneNumberFormat

class DataValidation extends ai.tripl.arc.plugins.UDFPlugin {

  val version = Utils.getFrameworkVersion

  // one udf plugin can register multiple user defined functions
  override def register()(implicit spark: SparkSession, logger: Logger, arcContext: ARCContext) = {

    // register custom UDFs via sqlContext.udf.register("funcName", func )
    spark.sqlContext.udf.register("is_valid_phonenumber", DataValidationPlugin.isValidPhoneNumber _ )
    spark.sqlContext.udf.register("format_phonenumber", DataValidationPlugin.formatPhoneNumber _ )
    spark.sqlContext.udf.register("is_valid_abn", DataValidationPlugin.isValidABN _ )

  }
}

object DataValidationPlugin {

  // uses google's libphonenumber to validate whether an input string phone number is valid for a specific region
  def isValidPhoneNumber(numberToParse: String, defaultRegion: String): Boolean = {
    try {
      val phoneUtil = PhoneNumberUtil.getInstance
      val parsed = phoneUtil.parse(numberToParse, defaultRegion)
      phoneUtil.isValidNumber(parsed)
    } catch {
      case e: Exception => false
    }
  }

  // uses google's libphonenumber to convert phone number to ISO E164 format https://en.wikipedia.org/wiki/E.164
  def formatPhoneNumber(numberToParse: String, defaultRegion: String): String = {
    val phoneUtil = PhoneNumberUtil.getInstance
    val parsed = phoneUtil.parse(numberToParse, defaultRegion)
    phoneUtil.format(parsed, PhoneNumberFormat.E164)
  }

  // validates whether an ABN passes the inbuilt checksum function
  // Subtract 1 from the first (left) digit to give a new eleven digit number.
  // Multiply each of the digits in this new number by its weighting factor.
  // Sum the resulting 11 products.
  // Divide the total by 89, noting the remainder.
  // If the remainder is zero the number is valid.
  def isValidABN(abn: String): Boolean = {
    // replace all non-digit characters
    val cleanABN = abn.replaceAll("\\D", "")

    // must be 11 characters
    if (cleanABN.length == 11) {
      val weightFactor = List(10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19)

      // calculate and verify checksum
      cleanABN.zipWithIndex.foldLeft(0) {
        case (sum, (digit, index)) => {
          // Subtract 1 from the first (left) digit
          val value = if (index == 0) digit.toString.toInt - 1 else digit.toString.toInt
          sum + (weightFactor(index) * value)
        }
      } % 89 == 0
    } else {
      false
    }
  }

}

src/main/scala/ai/tripl/arc/plugins/udf/DataQuality.scala
package ai.tripl.arc.plugins.udf

import scala.collection.JavaConverters._

import org.apache.spark.sql.SparkSession

import ai.tripl.arc.util.log.logger.Logger
import ai.tripl.arc.api.API.ARCContext
import ai.tripl.arc.util.Utils

import com.google.i18n.phonenumbers.PhoneNumberUtil
import com.google.i18n.phonenumbers.PhoneNumberUtil.PhoneNumberFormat

class DataQuality extends ai.tripl.arc.plugins.UDFPlugin {

  val version = ai.tripl.arc.plugins.udf.dataquality.BuildInfo.version

  // one udf plugin can register multiple user defined functions
  override def register()(implicit spark: SparkSession, logger: Logger, arcContext: ARCContext) = {

    // register custom UDFs via sqlContext.udf.register("funcName", func )
    spark.sqlContext.udf.register("is_valid_phonenumber", DataQualityPlugin.isValidPhoneNumber _ )
    spark.sqlContext.udf.register("format_phonenumber", DataQualityPlugin.formatPhoneNumber _ )
    spark.sqlContext.udf.register("is_valid_abn", DataQualityPlugin.isValidABN _ )
    spark.sqlContext.udf.register("is_valid_acn", DataQualityPlugin.isValidACN _ )

  }
}

object DataQualityPlugin {

  // uses google's libphonenumber to validate whether an input string phone number is valid for a specific region
  def isValidPhoneNumber(numberToParse: String, defaultRegion: String): Boolean = {
    try {
      val phoneUtil = PhoneNumberUtil.getInstance
      val parsed = phoneUtil.parse(numberToParse, defaultRegion)
      phoneUtil.isValidNumber(parsed)
    } catch {
      case e: Exception => false
    }
  }

  // uses google's libphonenumber to convert phone number to ISO E164 format https://en.wikipedia.org/wiki/E.164
  def formatPhoneNumber(numberToParse: String, defaultRegion: String): String = {
    val phoneUtil = PhoneNumberUtil.getInstance
    val parsed = phoneUtil.parse(numberToParse, defaultRegion)
    if (phoneUtil.isValidNumber(parsed)) {
      phoneUtil.format(parsed, PhoneNumberFormat.E164)
    } else {
      throw new Exception("Cannot format invalid phone number.")
    }
  }

  // validates whether an Australian Business Number (ABN) passes the inbuilt checksum function
  // Subtract 1 from the first (left) digit to give a new eleven digit number.
  // Multiply each of the digits in this new number by its weighting factor.
  // Sum the resulting 11 products.
  // Divide the total by 89, noting the remainder.
  // If the remainder is zero the number is valid.
  def isValidABN(abn: String): Boolean = {
    // replace all non-digit characters
    val cleanABN = abn.replaceAll("\\D", "")

    // must be 11 characters
    if (cleanABN.length == 11) {
      val weightFactor = List(10, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19)

      // calculate and verify checksum
      cleanABN.zipWithIndex.foldLeft(0) {
        case (sum, (digit, index)) => {
          // Subtract 1 from the first (left) digit
          val value = if (index == 0) digit.toString.toInt - 1 else digit.toString.toInt
          sum + (weightFactor(index) * value)
        }
      } % 89 == 0
    } else {
      false
    }
  }

  // validates whether an Australian Company Number (ACN) passes the inbuilt checksum function
  // Split the input ACN to remove the last digit which is the checksum digit
  // Multiply each of the remaining digits in this new number by its weighting factor.
  // Sum the resulting 8 products.
  // Divide the total by 10, noting the remainder.
  // if 10 - remainder = 10 then set to 0
  // if remainder = checksum digit then the number is valid.
  def isValidACN(acn: String): Boolean = {
    // replace all non-digit characters
    val cleanACN = acn.replaceAll("\\D", "")

    // must be 9 characters
    if (cleanACN.length == 9) {
      val weightFactor = List(8, 7, 6, 5, 4, 3, 2, 1)
      val (number, checksum) = cleanACN.splitAt(8)

      // calculate sum
      val sum = 10 - number.zipWithIndex.foldLeft(0) {
        case (sum, (digit, index)) => {
          sum + (weightFactor(index) * digit.toString.toInt)
        }
      } % 10 
      
      // validate sum
      if (sum == 10) checksum.toInt == 0 else checksum.toInt == sum
    } else {
      false
    }
  }   

}